lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,108 @@
+                if datetime.now().replace(tzinfo=UTC) > last_supported_date:
+                    return HttpResponse(status=426)  # Http status 426; Update Required
+
+    def process_response(self, __, response):
+        """
+        If request is from mobile native app, then add headers to response;
+        1. EDX-APP-LATEST-VERSION; if user app version < latest available version
+        2. EDX-APP-UPGRADE-DATE; if user app version < min supported version and timestamp < expiry of that version
+        """
+        request_cache_dict = request_cache.get_cache(self.REQUEST_CACHE_NAME)
+        if request_cache_dict:
+            last_supported_date = request_cache_dict[self.LAST_SUPPORTED_DATE_HEADER]
+            if last_supported_date != self.NO_LAST_SUPPORTED_DATE:
+                response[self.LAST_SUPPORTED_DATE_HEADER] = last_supported_date.isoformat()
+            latest_version = request_cache_dict[self.LATEST_VERSION_HEADER]
+            user_app_version = request_cache_dict[self.USER_APP_VERSION]
+            if (latest_version != self.NO_LATEST_VERSION and
+                    parsed_version(user_app_version) < parsed_version(latest_version)):
+                response[self.LATEST_VERSION_HEADER] = latest_version
+        return response
+
+    def _get_cache_key_name(self, field, key):
+        """ get key name to use to cache any property against field name and identification key """
+        return "mobile_api.app_version_upgrade.{}.{}".format(field, key)
+
+    def _get_version_info(self, request):
+        """
+        It sets request cache data for last_supported_date and latest_version with memcached values if exists against
+        user app properties else computes the values for specific platform and sets it in both memcache (for next
+        server interaction from same app version/platform) and request cache
+        """
+        user_agent = request.META.get('HTTP_USER_AGENT')
+        if user_agent:
+            platform = self._get_platform(request, user_agent)
+            if platform:
+                request_cache_dict = request_cache.get_cache(self.REQUEST_CACHE_NAME)
+                request_cache_dict[self.USER_APP_VERSION] = platform.version
+                last_supported_date_cache_key = self._get_cache_key_name(
+                    self.LAST_SUPPORTED_DATE_HEADER,
+                    platform.version
+                )
+                latest_version_cache_key = self._get_cache_key_name(self.LATEST_VERSION_HEADER, platform.NAME)
+                cached_data = cache.get_many([last_supported_date_cache_key, latest_version_cache_key])
+
+                last_supported_date = cached_data.get(last_supported_date_cache_key)
+                if not last_supported_date:
+                    last_supported_date = self._get_last_supported_date(platform.NAME, platform.version)
+                    cache.set(last_supported_date_cache_key, last_supported_date, self.CACHE_TIMEOUT)
```

Comment:
Where is the cached data updated/invalidated?  Shouldn't it be invalidated when the data in `AppVersionConfig` is updated?